File: Sources/Runtime/Networking/MelodyWebSocket.swift
import Foundation

/// Callback-driven WebSocket client used by Lua scripts for real-time communication.
final class MelodyWebSocket: NSObject, @unchecked Sendable, URLSessionWebSocketDelegate {
    private var task: URLSessionWebSocketTask?
    private var session: URLSession?

    var onOpen: (() -> Void)?
    var onMessage: ((String) -> Void)?
    var onError: ((String) -> Void)?
    var onClose: ((Int, String?) -> Void)?

    /// Connect to a WebSocket URL with optional headers.
    func connect(url: URL, headers: [String: String]?) {
        session = URLSession(
            configuration: .default,
            delegate: self,
            delegateQueue: nil
        )

        var request = URLRequest(url: url)
        headers?.forEach { key, value in
            request.setValue(value, forHTTPHeaderField: key)
        }

        task = session!.webSocketTask(with: request)
        task?.resume()
        listen()
    }

    /// Send a text message.
    func send(_ text: String) {
        task?.send(.string(text)) { error in
            if let error = error {
                DispatchQueue.main.async { [weak self] in
                    self?.onError?(error.localizedDescription)
                }
            }
        }
    }

    /// Close the connection with a code and optional reason.
    func close(code: URLSessionWebSocketTask.CloseCode = .normalClosure, reason: String? = nil) {
        let reasonData = reason?.data(using: .utf8)
        task?.cancel(with: code, reason: reasonData)
    }

    /// Force-close and release all callbacks.
    func disconnect() {
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
        session?.invalidateAndCancel()
        session = nil
        onOpen = nil
        onMessage = nil
        onError = nil
        onClose = nil
    }

    // MARK: - Private

    private func listen() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                DispatchQueue.main.async {
                    switch message {
                    case .string(let text):
                        self.onMessage?(text)
                    case .data(let data):
                        let text = String(data: data, encoding: .utf8) ?? ""
                        self.onMessage?(text)
                    @unknown default:
                        break
                    }
                }
                self.listen()
            case .failure(let error):
                DispatchQueue.main.async {
                    self.onError?(error.localizedDescription)
                }
            }
        }
    }

    // MARK: - URLSessionWebSocketDelegate

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol: String?
    ) {
        DispatchQueue.main.async { [weak self] in
            self?.onOpen?()
        }
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        let reasonString = reason.flatMap { String(data: $0, encoding: .utf8) }
        DispatchQueue.main.async { [weak self] in
            self?.onClose?(closeCode.rawValue, reasonString)
        }
    }

    // MARK: - URLSessionDelegate (SSL Trust)

    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
           let serverTrust = challenge.protectionSpace.serverTrust,
           MelodyURLSession.shared.isHostTrusted(challenge.protectionSpace.host) {
            completionHandler(.useCredential, URLCredential(trust: serverTrust))
        } else {
            completionHandler(.performDefaultHandling, nil)
        }
    }
}

File: Android/melody-runtime/src/main/java/com/melody/runtime/networking/MelodyWebSocket.kt
package com.melody.runtime.networking

import android.os.Handler
import android.os.Looper
import okhttp3.*

/**
 * WebSocket wrapper using OkHttp WebSocketListener.
 * Inherits SSL trust from MelodyHTTP's OkHttpClient.
 */
class MelodyWebSocket(
    var onOpen: (() -> Unit)? = null,
    var onMessage: ((String) -> Unit)? = null,
    var onError: ((String) -> Unit)? = null,
    var onClose: ((Int, String?) -> Unit)? = null
) {
    private var webSocket: WebSocket? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    /** Connect to a WebSocket URL with optional headers. */
    fun connect(url: String, headers: Map<String, String>?) {
        val requestBuilder = Request.Builder().url(url)
        headers?.forEach { (key, value) ->
            requestBuilder.addHeader(key, value)
        }

        val client = MelodyHTTP.getClient()
        webSocket = client.newWebSocket(requestBuilder.build(), object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                mainHandler.post { onOpen?.invoke() }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                mainHandler.post { onMessage?.invoke(text) }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                mainHandler.post { onError?.invoke(t.message ?: "WebSocket error") }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(code, reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                mainHandler.post { onClose?.invoke(code, reason.ifEmpty { null }) }
            }
        })
    }

    /** Send a text message. Returns true if enqueued successfully. */
    fun send(text: String): Boolean {
        return webSocket?.send(text) ?: false
    }

    /** Close the connection with a code and optional reason. */
    fun close(code: Int = 1000, reason: String? = null) {
        webSocket?.close(code, reason)
    }

    /** Force-close and release all callbacks. */
    fun disconnect() {
        webSocket?.cancel()
        webSocket = null
        onOpen = null
        onMessage = null
        onError = null
        onClose = null
    }
}
